addons/io_hubs_addon/third_party/recast.py

```diff
@@ -0,0 +1,636 @@
+        bound_box_x_coords.append(point.x)
+        bound_box_y_coords.append(point.y)
+
+    size_x = abs(min(bound_box_x_coords) - max(bound_box_x_coords))
+    size_y = abs(min(bound_box_y_coords) - max(bound_box_y_coords))
+    area = size_x * size_y
+
+    return pow(area, 1 / 3) / 50
+
+
+class RecastNavMeshResetOperator(bpy.types.Operator):
+    bl_idname = "recast.reset_navigation_mesh"
+    bl_label = "Reset"
+    bl_description = "Reset navigation mesh properties to default."
+    bl_options = {'REGISTER', 'UNDO'}
+
+    def execute(self, context):
+        scene = context.scene
+
+        scene.recast_navmesh.cell_size = CELL_SIZE_DEFAULT
+        scene.recast_navmesh.cell_height = CELL_HEIGHT_DEFAULT
+        scene.recast_navmesh.slope_max = SLOPE_MAX_DEFAULT
+        scene.recast_navmesh.climb_max = CLIMB_MAX_DEFAULT
+        scene.recast_navmesh.agent_height = AGENT_HEIGHT_DEFAULT
+        scene.recast_navmesh.agent_radius = AGENT_RADIUS_DEFAULT
+        scene.recast_navmesh.edge_max_len = EDGE_MAX_LENGTH
+        scene.recast_navmesh.edge_max_error = EDGE_MAX_ERROR
+        scene.recast_navmesh.region_min_size = REGION_MIN_SIZE
+        scene.recast_navmesh.region_merge_size = REGION_MERGE_SIZE
+        scene.recast_navmesh.verts_per_poly = VERTS_PER_POLY_DEFAULT
+        scene.recast_navmesh.sample_dist = SAMPLE_DIST_DEFAULT
+        scene.recast_navmesh.sample_max_error = SAMPLE_MAX_ERROR_DEFAULT
+        scene.recast_navmesh.partitioning = PARTITIONING_DEFAULT
+        scene.recast_navmesh.color = COLOR_DEFAULT
+        scene.recast_navmesh.auto_cell = AUTO_CELL_DEFAULT
+
+        return {'FINISHED'}
+
+
+class RecastNavMeshGenerateOperator(bpy.types.Operator):
+    bl_idname = "recast.build_navigation_mesh"
+    bl_label = "Build Navigation Mesh"
+    bl_description = "Build navigation mesh from the selected objects using recast."
+    bl_options = {'REGISTER', 'UNDO'}
+
+    def execute(self, context):
+        # bpy.ops.wm.call_menu(name="ADDITIVE_ANIMATION_insert_keyframe_menu")
+
+        selected_objects = context.selected_objects
+        if len(selected_objects) == 0:
+            self.report({'WARNING'}, 'No meshes selected')
+            return {"CANCELLED"}
+
+        from ..components.definitions.nav_mesh import NavMesh
+        nav_mesh_id = NavMesh.get_name()
+        for ob in context.selected_objects:
+            if has_component(ob, nav_mesh_id):
+                self.report({'ERROR'}, 'A Navmesh cannot be part of the selection')
+                return {'FINISHED'}
+
+        navMesh = None
+        navMeshes = get_objects_with_component(nav_mesh_id)
+        if navMeshes:
+            navMesh = navMeshes[0]
+
+        addon_prefs = get_addon_pref(context)
+        libpath = os.path.abspath(addon_prefs.recast_lib_path)
+        libpathr = libpath.replace("\\", "/")
+        if not os.path.exists(libpathr):
+            self.report({'ERROR'}, 'File not exists: %s\n' % libpathr)
+            return {'FINISHED'}
```

Comment:
I don't think any permanent changes have been made at this point, so I think this should return {'CANCELLED'} to avoid pushing an undo step.